Update serialization of <color> custom properties to used value

weinig opened the pull request:
#### 0fb43e3086742cde9ab118f669c8c9ba28ee0790
<pre>
Update serialization of &lt;color&gt; custom properties to used value
<a href="https://bugs.webkit.org/show_bug.cgi?id=294504">https://bugs.webkit.org/show_bug.cgi?id=294504</a>

Reviewed by Darin Adler and Tim Nguyen.

In <a href="https://github.com/w3c/csswg-drafts/issues/10371">https://github.com/w3c/csswg-drafts/issues/10371</a> it was resolved that:

```
RESOLVED: &lt;color&gt; values returned by getComputedStyle() on standard CSS properties or registered custom properties is the used value
```

By updating the implementation to match, we are able to remove quite a
lot of extra code in place only to serialize the computed color value,
including removing generic serialization for a few color types.

* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/registered-property-computation.html:
    - Update test for used value results. Set `color` property so that currentcolor tests are actually doing something.

* Source/WebCore/style/StyleCustomProperty.cpp:
    - Update to go through normal serialization path which takes a RenderStyle
      for computing the used color value.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/values/color/CSSAbsoluteColor.h:
* Source/WebCore/css/values/color/CSSAbsoluteColorSerialization.h: Removed.
* Source/WebCore/css/values/color/CSSColorLayers.cpp:
* Source/WebCore/css/values/color/CSSColorLayersSerialization.h: Removed.
* Source/WebCore/css/values/color/CSSColorMix.cpp:
* Source/WebCore/css/values/color/CSSColorMixSerialization.cpp: Removed.
* Source/WebCore/css/values/color/CSSColorMixSerialization.h: Removed.
* Source/WebCore/css/values/color/CSSContrastColor.cpp:
* Source/WebCore/css/values/color/CSSContrastColorSerialization.h: Removed.
* Source/WebCore/css/values/color/CSSRelativeColor.h:
* Source/WebCore/css/values/color/CSSRelativeColorSerialization.h: Removed.
    - Remove now unnecessary generic color type serialization files.

* Source/WebCore/style/values/color/StyleColor.cpp:
* Source/WebCore/style/values/color/StyleColor.h:
* Source/WebCore/style/values/color/StyleColorLayers.cpp:
* Source/WebCore/style/values/color/StyleColorLayers.h:
* Source/WebCore/style/values/color/StyleColorMix.cpp:
* Source/WebCore/style/values/color/StyleColorMix.h:
* Source/WebCore/style/values/color/StyleContrastColor.cpp:
* Source/WebCore/style/values/color/StyleContrastColor.h:
* Source/WebCore/style/values/color/StyleCurrentColor.cpp:
* Source/WebCore/style/values/color/StyleCurrentColor.h:
* Source/WebCore/style/values/color/StyleRelativeColor.h:
* Source/WebCore/style/values/color/StyleResolvedColor.cpp:
* Source/WebCore/style/values/color/StyleResolvedColor.h:
    - Remove serializationForCSS() overload that does not take a RenderStyle.
    - In a few places, updates the TextStream overloads to contain simplified
      serialization that used to just call serializationForCSS().

Canonical link: <a href="https://commits.webkit.org/296337@main">https://commits.webkit.org/296337@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/43a56fa353e8c403b0e18aa6e358e0d738efcf47

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/108197 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/27860 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/18280 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/113408 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/58682 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/110160 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/28560 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/36411 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/82154 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/58682 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/111145 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/22629 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/97466 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/62585 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/22047 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/15602 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/58144 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/91995 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/15658 "Found 1 new test failure: imported/w3c/web-platform-tests/feature-policy/feature-policy-frame-policy-timing.https.sub.html (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/116534 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/35263 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/25965 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/91182 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/35637 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/93742 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/90977 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23186 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/35864 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13628 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/31042 "Build was cancelled. Recent messages:Printed configuration") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/35162 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/40718 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/34896 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38253 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/36563 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m reviewed (approved):
I assume this is just code moved over.

Please don't forget to export the WPT change :)

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/registered-property-computation.html:
```diff
@@ -154,15 +155,15 @@
 test_computed_value('<color>', 'tomato', 'rgb(255, 99, 71)');
 test_computed_value('<color>', 'plum', 'rgb(221, 160, 221)');
-test_computed_value('<color>', 'currentcolor', 'currentcolor');
+test_computed_value('<color>', 'currentcolor', 'rgb(0, 0, 255)');
```
Do we have tests for the specified values of custom properties somewhere? I think we'd still want to test that this still returns currentColor for the specified value. Ditto for the testcases below.

darinadler replied:
I don’t think we do, but such tests seem like they would fit in with the other tests in the registered-property-cssom.html file in this same directory in WPT.

darinadler commented on Source/WebCore/css/values/color/CSSColorMix.cpp:
```diff
@@ -90,9 +90,95 @@ bool containsColorSchemeDependentColor(const ColorMix& unresolved)
+    return WTF::switchOn(percentage,
+        [](const ColorMix::Component::Percentage::Raw& raw) { return raw.value == 50.0; },
+        [](const ColorMix::Component::Percentage::Calc&) { return false; }
+    );
```
Given the style of the function just below, maybe this instead?
```suggestion
    return !percentage.isCalc() && percentage.raw()->value == 50.0;
```
Or maybe we want to use `switchOn` except when there are multiple objects involved?

weinig replied:
I can actually do this one so much better.

```
return percentage == 50_css_percentage;
```

Boom.

darinadler commented on Source/WebCore/css/values/color/CSSRelativeColor.h:
```diff
@@ -100,7 +101,32 @@ bool containsCurrentColor(const RelativeColor<Descriptor>& unresolved)
+            builder.append("color(from "_s);
+            serializationForCSS(builder, context, value.origin);
+            builder.append(' ');
+            builder.append(serialization(ColorSpaceFor<typename D::ColorType>));
```
```suggestion
            builder.append(' ', serialization(ColorSpaceFor<typename D::ColorType>));
```

darinadler commented on Source/WebCore/css/values/color/CSSRelativeColor.h:
```diff
@@ -100,7 +101,32 @@ bool containsCurrentColor(const RelativeColor<Descriptor>& unresolved)
+            builder.append(serialization(ColorSpaceFor<typename D::ColorType>));
+        } else {
+            builder.append(D::serializationFunctionName);
+            builder.append("(from "_s);
```
```suggestion
            builder.append(D::serializationFunctionName, "(from "_s);
```

darinadler commented on Source/WebCore/style/values/color/StyleColorLayers.cpp:
```diff
@@ -92,25 +92,18 @@ bool containsCurrentColor(const ColorLayers& colorLayers)
+    if (colorLayers.blendMode != BlendMode::Normal)
+        ts << colorLayers.blendMode << ", "_s;
+    ts << colorLayers.colors;
+    ts << ')';
```
```suggestion
    ts << colorLayers.colors << ')';
```
Not sure this is as important as in `StringBuilder::append`, where doing multiple in a single call is more efficient, but it does at least seem idiomatic to combine into a single line when possible.

weinig committed:
Update serialization of <color> custom properties to used value
https://bugs.webkit.org/show_bug.cgi?id=294504

Reviewed by Darin Adler and Tim Nguyen.

In https://github.com/w3c/csswg-drafts/issues/10371 it was resolved that:

```
RESOLVED: <color> values returned by getComputedStyle() on standard CSS properties or registered custom properties is the used value
```

By updating the implementation to match, we are able to remove quite a
lot of extra code in place only to serialize the computed color value,
including removing generic serialization for a few color types.

* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/registered-property-computation.html:
    - Update test for used value results. Set `color` property so that currentcolor tests are actually doing something.

* Source/WebCore/style/StyleCustomProperty.cpp:
    - Update to go through normal serialization path which takes a RenderStyle
      for computing the used color value.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/values/color/CSSAbsoluteColor.h:
* Source/WebCore/css/values/color/CSSAbsoluteColorSerialization.h: Removed.
* Source/WebCore/css/values/color/CSSColorLayers.cpp:
* Source/WebCore/css/values/color/CSSColorLayersSerialization.h: Removed.
* Source/WebCore/css/values/color/CSSColorMix.cpp:
* Source/WebCore/css/values/color/CSSColorMixSerialization.cpp: Removed.
* Source/WebCore/css/values/color/CSSColorMixSerialization.h: Removed.
* Source/WebCore/css/values/color/CSSContrastColor.cpp:
* Source/WebCore/css/values/color/CSSContrastColorSerialization.h: Removed.
* Source/WebCore/css/values/color/CSSRelativeColor.h:
* Source/WebCore/css/values/color/CSSRelativeColorSerialization.h: Removed.
    - Remove now unnecessary generic color type serialization files.

* Source/WebCore/style/values/color/StyleColor.cpp:
* Source/WebCore/style/values/color/StyleColor.h:
* Source/WebCore/style/values/color/StyleColorLayers.cpp:
* Source/WebCore/style/values/color/StyleColorLayers.h:
* Source/WebCore/style/values/color/StyleColorMix.cpp:
* Source/WebCore/style/values/color/StyleColorMix.h:
* Source/WebCore/style/values/color/StyleContrastColor.cpp:
* Source/WebCore/style/values/color/StyleContrastColor.h:
* Source/WebCore/style/values/color/StyleCurrentColor.cpp:
* Source/WebCore/style/values/color/StyleCurrentColor.h:
* Source/WebCore/style/values/color/StyleRelativeColor.h:
* Source/WebCore/style/values/color/StyleResolvedColor.cpp:
* Source/WebCore/style/values/color/StyleResolvedColor.h:
    - Remove serializationForCSS() overload that does not take a RenderStyle.
    - In a few places, updates the TextStream overloads to contain simplified
      serialization that used to just call serializationForCSS().

Canonical link: https://commits.webkit.org/296337@main
